schism/vgamem.c

```diff
@@ -768,70 +799,121 @@ void draw_vu_meter(int x, int y, int width, int val, int color, int peak)
 
 /* --------------------------------------------------------------------- */
 /* sample drawing
- * 
+ *
  * output channels = number of oscis
  * input channels = number of channels in data
-*/
+ */
 
-/* somewhat heavily based on CViewSample::DrawSampleData2 in modplug
- *
- * TODO: vectorized sample drawing. */
-#define DRAW_SAMPLE_DATA_VARIANT(bits, doublebits) \
-	static void _draw_sample_data_##bits(struct vgamem_overlay *r, \
-		int##bits##_t *data, uint32_t length, unsigned int inputchans, unsigned int outputchans) \
+#define DRAW_SAMPLE_DATA_EX_VARIANT(BITS, SCALE, SIGNARG_DEF, SIGN_FLIP_DEF, SIGN_FLIP_DO) \
+	void draw_sample_data_ex_##BITS(struct vgamem_overlay *r, int x1, int y1, int x2, int y2, \
+		int##BITS##_t *data, int offset, int stride, int advance, int length, SIGNARG_DEF \
```

Comment:
can't `advance` be calculated with all the info already provided?

p.s. `offset`, `stride`, and `length` should all be `int32_t` or `uint32_t`

Reply:
`advance` is calculated just before the call to `draw_sample_data_ex`:
```
		int advance = _muldiv(sample->length * chans, 256, r->width * stride);
```
I don't think that information is available in the current parameter set. But, it's also a forced-inline function, and in the current implementation, it's only calculated in one place.

Reply:
Oh, also, `draw_data_sample_rect_BITS` also uses `draw_sample_data_ex_BITS` and uses a different `advance` calculation.